logs/stream/stream.go

```diff
@@ -151,14 +176,28 @@ func AnalyzeStreamInGroups(logLines []state.LogLine) (state.TransientLogState, s
 	//
 	// Its important we do this early, to support out-of-order receipt of log lines,
 	// up to the freshness threshold used in the next function call (3 seconds)
+	allLinesHaveBackendPid := true
+	allLinesHaveLogLineNumber := true
+	allLinesHaveOccurredAt := true
+	for _, logLine := range logLines {
+		if logLine.BackendPid == 0 {
+			allLinesHaveBackendPid = false
+		}
+		if logLine.LogLineNumber == 0 {
+			allLinesHaveLogLineNumber = false
+		}
+		if logLine.OccurredAt.IsZero() {
+			allLinesHaveOccurredAt = false
+		}
+	}
 	sort.SliceStable(logLines, func(i, j int) bool {
-		if logLines[i].BackendPid != 0 && logLines[j].BackendPid != 0 && logLines[i].BackendPid != logLines[j].BackendPid {
+		if allLinesHaveBackendPid && logLines[i].BackendPid != logLines[j].BackendPid {
 			return logLines[i].BackendPid < logLines[j].BackendPid
 		}
-		if logLines[i].LogLineNumber != 0 && logLines[j].LogLineNumber != 0 && logLines[i].LogLineNumber != logLines[j].LogLineNumber {
+		if allLinesHaveLogLineNumber && logLines[i].LogLineNumber != logLines[j].LogLineNumber {
 			return logLines[i].LogLineNumber < logLines[j].LogLineNumber
 		}
-		if !logLines[i].OccurredAt.IsZero() && !logLines[j].OccurredAt.IsZero() {
+		if allLinesHaveOccurredAt {
 			return logLines[i].OccurredAt.Sub(logLines[j].OccurredAt) < 0
 		}
 		return false // Keep initial order
```

Comment:
So we keep initial order if at least one line is missing each of the above, right? Basically, "sort by pid if all lines have it, then sort by line number if all lines have it, then sort by occurred_at if all lines have it, otherwise keep original sort order"? My naïve approach would be "sort by pids if both lines have pids, then sort by line number if both lines have line numbers, then sort by occurred_at if both lines have occurred_at, otherwise keep original sort order", but I'm guessing that doesn't do the right thing somehow? (Since it's simpler, I imagine you considered it.)

Reply:
@uhoh-itsmaciek If I read your comment right, thats actually exactly the logic thats being replaced here. i.e. before we were only checking on a per-line basis whether both lines had a value (and then we sorted by it), and now we only sort by it if all the lines have it.

The problem with sorting by e.g. PID for some lines, but not others, is that multi-line statements are not handled correctly. e.g. in the self-managed case, EXPLAIN output would usually be multi-line, with only the first line having a PID. Thus, that line would be compared with other lines (and moved around), but everything in the multi-line statement that came after it, would not be sorted. Thus the output is buggy